File: ios/SampleUI/Sources/SampleUI/Components/UseSmileIDSampleSessionCard.swift
import SwiftUI

/// The active session and its countdown; `remaining` arrives formatted, the deadline being absolute.
public struct UseSmileIDSampleSessionCard: View {
  private let sessionId: String
  private let remaining: String

  @ScaledMetric(relativeTo: .body) private var minHeight: CGFloat = SmileSpacing.space64
  @Environment(\.useSmileIDSampleColors) private var colors

  public init(sessionId: String, remaining: String) {
    self.sessionId = sessionId
    self.remaining = remaining
  }

  public var body: some View {
    HStack(spacing: SmileSpacing.spacingSm) {
      VStack(alignment: .leading, spacing: SmileSpacing.spacingXxs) {
        UseSmileIDSampleText(
          "ACTIVE TOKEN SESSION",
          style: UseSmileIDSampleTheme.type.textStyleOverline.with(tracking: smileLabelTracking)
        )
        UseSmileIDSampleText("Linked to session \(sessionId)", style: UseSmileIDSampleTheme.type.textStyleCaption)
      }
      .frame(maxWidth: .infinity, alignment: .leading)
      // The one value that must stay whole, so the text beside it yields instead.
      UseSmileIDSampleText(remaining, style: UseSmileIDSampleTheme.type.textStyleHeadingCard.with(size: 20))
        .fixedSize()
        .useSmileIDSampleTestId(UseSmileIDSampleTestIds.sessionCountdown)
    }
    // The gradient is scheme-independent, so its ink is too.
    .foregroundColor(SmileColorLight.colorTextInverse)
    .padding(SmileSpacing.spacingMd)
    .frame(minHeight: minHeight)
    .background(
      LinearGradient(
        colors: zip(smileTokenSessionGradient, smileTokenSessionGradientAlpha).map { $0.opacity($1) },
        startPoint: .leading,
        endPoint: .trailing
      )
    )
    .clipShape(RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous)
        .strokeBorder(colors.cardStroke, lineWidth: smileCardStrokeWidth)
    )
    // A container that contains, or the card's id replaces the countdown's and a flow cannot read it.
    .accessibilityElement(children: .contain)
    .useSmileIDSampleTestId(UseSmileIDSampleTestIds.sessionCard)
  }
}

/// Replaces ``UseSmileIDSampleSessionCard`` on expiry: a neutral card, not a warning-accented one.
public struct UseSmileIDSampleSessionEndedBanner: View {
  private let onScan: () -> Void

  @ScaledMetric(relativeTo: .body) private var minHeight: CGFloat = SmileSpacing.space64
  @Environment(\.useSmileIDSampleColors) private var colors

  public init(onScan: @escaping () -> Void) {
    self.onScan = onScan
  }

  public var body: some View {
    HStack(spacing: SmileSpacing.spacingSm) {
      VStack(alignment: .leading, spacing: SmileSpacing.spacingXxs) {
        UseSmileIDSampleText("TOKEN SESSION ENDED", style: UseSmileIDSampleTheme.type.textStyleOverline)
          .foregroundColor(colors.banner.text)
        UseSmileIDSampleText("Scan a token to relink", style: UseSmileIDSampleTheme.type.textStyleBodyStrong)
          .foregroundColor(colors.banner.title)
      }
      .frame(maxWidth: .infinity, alignment: .leading)
      // A text action widened to the touch target, per the component's spec — not a filled pill.
      Button(action: onScan) {
        UseSmileIDSampleText("Scan", style: UseSmileIDSampleTheme.type.linkFont)
          .foregroundColor(colors.primary)
          .fixedSize()
          .padding(.horizontal, SmileSpacing.spacingXs)
          .frame(minHeight: SmileSpacing.sizeControlMd)
          .contentShape(Rectangle())
      }
      .buttonStyle(.plain)
    }
    .padding(SmileSpacing.spacingMd)
    .frame(minHeight: minHeight)
    // Surface-muted per this component's token list; the banner contract's fill is a warm sand.
    .background(colors.surfaceMuted)
    .clipShape(RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous)
        .strokeBorder(colors.cardStroke, lineWidth: smileCardStrokeWidth)
    )
    .useSmileIDSampleTestId(UseSmileIDSampleTestIds.sessionEndedBanner)
  }
}

File: flutter/sample_ui/lib/src/components/use_smileid_sample_session_card.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';
import '../use_smileid_sample_test_ids.dart';

/// The active token session and its m:ss countdown.
///
/// [remaining] arrives formatted, because the deadline is absolute and the ticking is the screen's.
/// Both gradient stops are translucent, so the card composites against the page and reads
/// differently per scheme by design.
class UseSmileIDSampleSessionCard extends StatelessWidget {
  /// Takes the formatted countdown rather than a duration, so no component holds a clock.
  const UseSmileIDSampleSessionCard({
    required this.sessionId,
    required this.remaining,
    super.key,
  });

  /// The linked session's id.
  final String sessionId;

  /// The countdown, already formatted as m:ss.
  final String remaining;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    // The gradient is scheme-independent, so its ink is too: color.surface is #272A35 in dark.
    final Color ink = SmileColorLight.colorTextInverse;
    return Semantics(
      identifier: UseSmileIDSampleTestIds.sessionCard,
      container: true,
      child: _SessionSurface(
        decoration: BoxDecoration(
          borderRadius: UseSmileIDSampleShapes.card,
          gradient: LinearGradient(colors: _sessionGradient),
          border: Border.all(
            color: colors.cardStroke,
            width: smileCardStrokeWidth,
          ),
        ),
        lead: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text(
              'ACTIVE TOKEN SESSION',
              style: UseSmileIDSampleType.textStyleOverline.copyWith(
                letterSpacing: _labelTracking,
                color: ink,
              ),
            ),
            const SizedBox(height: SmileDimens.spacingXxs),
            Text(
              'Linked to session $sessionId',
              style: UseSmileIDSampleType.textStyleCaption.copyWith(
                fontWeight: FontWeight.values[smileCardFamilyWeight ~/ 100 - 1],
                color: ink,
              ),
            ),
          ],
        ),
        trail: Semantics(
          identifier: UseSmileIDSampleTestIds.sessionCountdown,
          child: Text(
            remaining,
            // The one value on this card that must stay whole; the text beside it yields.
            softWrap: false,
            style: UseSmileIDSampleType.textStyleHeadingCard.copyWith(
              fontSize: _countdownSize,
              color: ink,
            ),
          ),
        ),
      ),
    );
  }
}

/// Replaces the session card on expiry: a neutral card, not a warning-accented one.
class UseSmileIDSampleSessionEndedBanner extends StatelessWidget {
  /// Matches the session card's height, so the two swap without the layout moving.
  const UseSmileIDSampleSessionEndedBanner({required this.onScan, super.key});

  /// Reaches the scan screen to relink.
  final VoidCallback onScan;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Semantics(
      identifier: UseSmileIDSampleTestIds.sessionEndedBanner,
      container: true,
      child: _SessionSurface(
        decoration: BoxDecoration(
          borderRadius: UseSmileIDSampleShapes.card,
          // surface-muted per this component's token list; the banner fill is a warm sand.
          color: colors.surfaceMuted,
          border: Border.all(
            color: colors.cardStroke,
            width: smileCardStrokeWidth,
          ),
        ),
        lead: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text(
              'TOKEN SESSION ENDED',
              style: UseSmileIDSampleType.textStyleOverline.copyWith(
                color: colors.banner.text,
              ),
            ),
            const SizedBox(height: SmileDimens.spacingXxs),
            Text(
              'Scan a token to relink',
              style: UseSmileIDSampleType.textStyleBodyStrong.copyWith(
                color: colors.banner.title,
              ),
            ),
          ],
        ),
        trail: Semantics(
          button: true,
          child: GestureDetector(
            onTap: onScan,
            behavior: HitTestBehavior.opaque,
            child: ConstrainedBox(
              constraints: const BoxConstraints(
                minWidth: SmileDimens.sizeControlMd,
                minHeight: SmileDimens.sizeControlMd,
              ),
              child: Padding(
                padding: const EdgeInsets.symmetric(
                  horizontal: SmileDimens.spacingXs,
                ),
                child: Center(
                  child: Text(
                    'Scan',
                    softWrap: false,
                    style: UseSmileIDSampleType.linkFont.copyWith(
                      color: colors.primary,
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

/// The shape both cards take, so the two swap without the layout moving.
///
/// A flexible leading column rather than a wrap: the trailing half keeps its intrinsic width and
/// the text beside it wraps, so the card grows instead of the action moving under the text. A wrap
/// would need to be told what the trailing half costs, and guessed that wrong on both cards.
class _SessionSurface extends StatelessWidget {
  const _SessionSurface({
    required this.decoration,
    required this.lead,
    required this.trail,
  });

  final BoxDecoration decoration;
  final Widget lead;
  final Widget trail;

  @override
  Widget build(BuildContext context) => DecoratedBox(
    decoration: decoration,
    child: ConstrainedBox(
      constraints: const BoxConstraints(minHeight: SmileDimens.space64),
      child: Padding(
        padding: const EdgeInsets.all(SmileDimens.spacingMd),
        child: Row(
          children: <Widget>[
            Expanded(child: lead),
            const SizedBox(width: SmileDimens.spacingSm),
            trail,
          ],
        ),
      ),
    ),
  );
}

/// Alpha is applied here, never baked into the token's hex.
final List<Color> _sessionGradient = <Color>[
  for (int index = 0; index < smileTokenSessionGradient.length; index++)
    smileTokenSessionGradient[index].withValues(
      alpha: smileTokenSessionGradientAlpha[index],
    ),
];

/// The overline's own tracking on this card, which the label style does not carry.
const double _labelTracking = 1;

/// The countdown's run.
const double _countdownSize = 24;
